pkg/client/metrics/metrics.go

```diff
@@ -69,20 +87,24 @@ type instrumentedGenericClient struct {
 
 func (c *instrumentedGenericClient) Do(ctx context.Context, verb, endpoint string, query url.Values) (client.APIResponse, error) {
 	startTime := time.Now()
+
 	var err error
+	var resp client.APIResponse
 	defer func() {
 		endTime := time.Now()
-		// skip calls where we don't make the actual request
 		if err != nil {
-			if _, wasAPIErr := err.(*client.Error); !wasAPIErr {
-				// TODO: measure API errors by code?
-				return
+			if apiErr, wasAPIErr := err.(*client.Error); wasAPIErr {
+				apiRequestsTotal.With(prometheus.Labels{"code": string(apiErr.Type), "path": endpoint, "server": c.serverName}).Inc()
+			} else {
+				apiRequestsTotal.With(prometheus.Labels{"code": "generic", "path": endpoint, "server": c.serverName}).Inc()
 			}
```

Comment:
When I first looked at your PR I had forgotten that the client errors were wrapped https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/pkg/client/types.go#L27-L31 which is not ideal as it is less explicit to users than a normal http code would be. I feel like we should change some things in order to surface the http code and expose it in the metrics.

I'll be out till the 27th of May, but I'll try to have a look at that again when I am back.

Reply:
@dgrisonnet  I updated code to add HTTPStatusCode to APIResponse to expose the http code in the metric. Can you please review? Thank you.